In [2]:
# The first step of using numpy is to tell python to import it
import numpy as np

### 2.1 NumPy Arrays

**NumPy Array**
* An array is a data structure that stores values of same data type.
* While python lists can contain values corresponding to different data types, arrays in python can only contain values corresponding to the same data type.
* However python lists fail to deliver the performance required while computing large sets of numerical data. To address this issue we use NumPy arrays.
* We can create NumPy arrays by converting a list to an array.


In [3]:
# defining a list of different car companies or string elements
arr_str = ['Mercedes', 'BMW', 'Audi', 'Ferrari', 'Tesla']

# defining a list of number of cylinders in car or numerical elements
arr_num = [5, 4, 6, 7, 3]

In [4]:
# connverting the list arr_str to a NumPy array
np_arr_str = np.array(arr_str)


# connverting the list arr_num to a NumPy array
np_arr_num = np.array(arr_num)

# checking the output
print('Numpy Array (arr_str): ',np_arr_str)
print('Numpy Array (arr_num): ',np_arr_num)

Numpy Array (arr_str):  ['Mercedes' 'BMW' 'Audi' 'Ferrari' 'Tesla']
Numpy Array (arr_num):  [5 4 6 7 3]


The resuts look similar to a list but arr_str and arr_num have been converted to NumPy arrays. Let's check the data type to confirm this.

In [6]:
# printing the data type of lists
print('Data type of arr_str: ',type(arr_str))
print('Data type of arr_num: ',type(arr_num))

# printing the data type after conversion of lists to array
print('Data type of np_arr_str: ',type(np_arr_str))
print('Data type of np_arr_num: ',type(np_arr_num))

Data type of arr_str:  <class 'list'>
Data type of arr_num:  <class 'list'>
Data type of np_arr_str:  <class 'numpy.ndarray'>
Data type of np_arr_num:  <class 'numpy.ndarray'>


* The above output confirms that both the lists were successfully converted to arrays

**NumPy Matrix**

* A matrix is a two-dimensional data structure where elements are arranged into rows and columns.
* A matrix can be created by using list of lists

In [8]:
# let's say we have information of different number of cylinders in a car and we want to display them in a matrix format
matrix = np.array([[1,2,1],[4,5,9],[1,8,9]])
print(matrix)

[[1 2 1]
 [4 5 9]
 [1 8 9]]


In [9]:
print('Data type of matrix: ',type(matrix))

Data type of matrix:  <class 'numpy.ndarray'>


* We see that all the NumPy objects have data type as ndarray

### 2.2 NumPy Functions

**There are different ways to create NumPy arrays using the functions available in NumPy library**

**Using np.arange() function**
* The np.arange() function returns an array with evenly spaced elements as per the interval. The interval mentioned is half-opened i.e. start is included but stop is excluded.
* It has the following paramaters:
  * start : start of interval range. By default start = 0
  * stop  : end of interval range
  * step  : step size of interval. By default step size = 1

In [10]:
arr2  = np.arange(start = 0, stop = 10) # 10 will be excluded from the output
print(arr2)

# or

arr2  = np.arange(0,10)
print(arr2)

[0 1 2 3 4 5 6 7 8 9]
[0 1 2 3 4 5 6 7 8 9]


In [11]:
# adding a step size of 5 to create an array
arr3  = np.arange(start = 0, stop = 20, step = 5)
arr3

array([ 0,  5, 10, 15])

**Using np.linspace() function**
* The np.linspace() function returns numbers which are evenly distributed with respect to interval. Here the start and stop both are included.            
*It has the following parameters:              
 * start: start of interval range. By default start = 0
 * stop: end of interval range
 * num : No. of samples to generate. By default num = 50

In [12]:
matrix2 = np.linspace(0,5) # by default 50 evenly spaced values will be generated between 0 and 5
matrix2

array([0.        , 0.10204082, 0.20408163, 0.30612245, 0.40816327,
       0.51020408, 0.6122449 , 0.71428571, 0.81632653, 0.91836735,
       1.02040816, 1.12244898, 1.2244898 , 1.32653061, 1.42857143,
       1.53061224, 1.63265306, 1.73469388, 1.83673469, 1.93877551,
       2.04081633, 2.14285714, 2.24489796, 2.34693878, 2.44897959,
       2.55102041, 2.65306122, 2.75510204, 2.85714286, 2.95918367,
       3.06122449, 3.16326531, 3.26530612, 3.36734694, 3.46938776,
       3.57142857, 3.67346939, 3.7755102 , 3.87755102, 3.97959184,
       4.08163265, 4.18367347, 4.28571429, 4.3877551 , 4.48979592,
       4.59183673, 4.69387755, 4.79591837, 4.89795918, 5.        ])

**How are these values getting generated?**

The step size or the difference between each element will be decided by the following formula:

**(stop - start) / (total elements - 1)**

So, in this case:
(5 - 0) / 49 = 0.10204082

The first value will be 0.10204082, the second value will be 0.10204082 + 0.10204082, the third value will be 0.10204082 + 0.10204082 +0.10204082, and so on.

In [13]:
# generating 10 evenly spaced values between 10 and 20
matrix3 = np.linspace(10,20,10)
matrix3

array([10.        , 11.11111111, 12.22222222, 13.33333333, 14.44444444,
       15.55555556, 16.66666667, 17.77777778, 18.88888889, 20.        ])

**Similarly we can create matrices using the functions available in NumPy library**

**Using np.zeros()**

* The np.zeros() is a function for creating a matrix and performing matrix operations in NumPy.
* It returns a matrix filled with zeros of the given shape.
* It has the following parameters:    
  * shape : Number of rows and columns in the output matrix.
  * dtype: data type of the elements in the matrix, by default the value is set to `float`.

In [16]:
matrix4 = np.zeros([3,5], int)
matrix4

array([[0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0]])

**Using np.ones()**

* The np.ones() is another function for creating a matrix and performing matrix operations in NumPy.
* It returns a matrix of given shape and type, filled with ones.
* It has the following parameters:  
  * shape : Number of rows and columns in the output matrix.
  * dtype: data type of the elements in the matrix, by default the value is set to `float`.

In [15]:
matrix5 = np.ones([3,5])
matrix5

array([[1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1.]])

**Using np.eye()**
* The np.eye() is a function for creating a matrix and performing matrix operations in NumPy.
* It returns a matrix with ones on the diagonal and zeros elsewhere.
* It has the following parameters:
  * n: Number of rows and columns in the output matrix
  * dtype: data type of the elements in the matrix, by default the value is set to `float`.

In [17]:
matrix6 = np.eye(5)
matrix6

array([[1., 0., 0., 0., 0.],
       [0., 1., 0., 0., 0.],
       [0., 0., 1., 0., 0.],
       [0., 0., 0., 1., 0.],
       [0., 0., 0., 0., 1.]])

**We can also convert a one dimension array to a matrix. This can be done by using the np.reshape() function.**

* The shape of an array basically tells the number of elements and dimensions of the array. Reshaping a Numpy array simply means changing the shape of the given array.
* By reshaping an array we can add or remove dimensions or change number of elements in each dimension.
* In order to reshape a NumPy array, we use the reshape method with the given array.
* **Syntax:** array.reshape(shape)
  * shape: a tuple given as input, the values in tuple will be the new shape of the array.

In [18]:
# defining an array with values 0 to 9
arr4 = np.arange(0,10)
arr4

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

In [19]:
# reshaping the array arr4 to a 2 x 5 matrix
arr4_reshaped = arr4.reshape((2,5))
arr4_reshaped

array([[0, 1, 2, 3, 4],
       [5, 6, 7, 8, 9]])

In [20]:
arr4

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

In [21]:
# reshaping the array arr4 to a 2 x 6 matrix
arr4.reshape((2,6))

ValueError: cannot reshape array of size 10 into shape (2,6)

In [22]:
len(arr4)

10

* This did not work because we have 10 elements which we are trying to fit in a 2 X 6 shape which will require 12 elements.

**NumPy can also perform a large number of different mathematical operations and it provides different functions to do so.**

NumPy provides:
1. Trigonometric functions
2. Exponents and Logarithmic functions
3. Functions for arithmetic operations between arrays and matrices

**Trigonometric functions**

In [23]:
print('Sine Function:',np.sin(4))
print('Cosine Function:',np.cos(4))
print('Tan Function',np.tan(4))

Sine Function: -0.7568024953079282
Cosine Function: -0.6536436208636119
Tan Function 1.1578212823495775


**Exponents and Logarithmic functions**

* Exponents

In [24]:
np.exp(2)

7.38905609893065

In [25]:
arr5 = np.array([2,4,6])
np.exp(arr5)

array([  7.3890561 ,  54.59815003, 403.42879349])

* Logarithms

In [26]:
# by default NumPy takes the base of log as e
np.log(2)

0.6931471805599453

In [27]:
np.log(arr5)

array([0.69314718, 1.38629436, 1.79175947])

In [28]:
## log with base 10
np.log10(8)

0.9030899869919435

**Arithmetic Operations on arrays**

In [29]:
# arithmetic on lists

l1 = [1,2,3]
l2 = [4,5,6]
print(l1+l2)
# this does not behave as you would expect!


[1, 2, 3, 4, 5, 6]


In [30]:
# we can +-*/ arrays together

# defining two arrays
arr7 = np.arange(1,6)
print('arr7:', arr7)

arr8 = np.arange(3,8)
print('arr8:', arr8)

arr7: [1 2 3 4 5]
arr8: [3 4 5 6 7]


In [ ]:
print('Addition: ',arr7+arr8)
print('Subtraction: ',arr8-arr7)
print('Multiplication:' , arr7*arr8)
print('Division:', arr7/arr8)
print('Inverse:', 1/arr7)
print('Powers:', arr7**arr8) # in python, powers are achieved using **, NOT ^!!! ^ does something completely different!

Addition:  [ 4  6  8 10 12]
Subtraction:  [2 2 2 2 2]
Multiplication: [ 3  8 15 24 35]
Division: [0.33333333 0.5        0.6        0.66666667 0.71428571]
Inverse: [1.         0.5        0.33333333 0.25       0.2       ]
Powers: [    1    16   243  4096 78125]


**Operations on Matrices**

In [31]:
matrix7 = np.arange(1,10).reshape(3,3)
print(matrix7)

matrix8 = np.eye(3)
print(matrix8)

[[1 2 3]
 [4 5 6]
 [7 8 9]]
[[1. 0. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]


In [32]:
print('Addition: \n', matrix7+matrix8)
print('Subtraction: \n ', matrix7-matrix8)
print('Multiplication: \n', matrix7*matrix8)
print('Division: \n', matrix7/matrix8)

Addition: 
 [[ 2.  2.  3.]
 [ 4.  6.  6.]
 [ 7.  8. 10.]]
Subtraction: 
  [[0. 2. 3.]
 [4. 4. 6.]
 [7. 8. 8.]]
Multiplication: 
 [[1. 0. 0.]
 [0. 5. 0.]
 [0. 0. 9.]]
Division: 
 [[ 1. inf inf]
 [inf  5. inf]
 [inf inf  9.]]


<ipython-input-32-b49c47b4c85a>:4: RuntimeWarning: divide by zero encountered in divide
  print('Division: \n', matrix7/matrix8)


* RuntimeWarning: Errors which occur during program execution(run-time) after successful compilation are called run-time errors.
* One of the most common run-time error is division by zero also known as Division error.
* Due to division by zero error, we are getting inf (infinity) values because 1/0 is not a defined operation.

**Linear algebra matrix multiplication**

In [33]:
matrix9 = np.arange(1,10).reshape(3,3)
print('First Matrix: \n',matrix9)

matrix10 = np.arange(11,20).reshape(3,3)
print('Second Matrix: \n',matrix10)
print('')
# taking linear algebra matrix multiplication (some may have heard this called the dot product)
print('Multiplication: \n', matrix9 @ matrix10)

First Matrix: 
 [[1 2 3]
 [4 5 6]
 [7 8 9]]
Second Matrix: 
 [[11 12 13]
 [14 15 16]
 [17 18 19]]

Multiplication: 
 [[ 90  96 102]
 [216 231 246]
 [342 366 390]]


In [34]:
[1,2,3]
[11, 14, 17]

[11, 14, 17]

In [36]:
1 * 11 +2 * 14 +3 *17

90

**Transpose of a matrix**

In [37]:
print(matrix9)

[[1 2 3]
 [4 5 6]
 [7 8 9]]


In [38]:
# taking transpose of matrix
np.transpose(matrix9)

array([[1, 4, 7],
       [2, 5, 8],
       [3, 6, 9]])

In [39]:
# another way of taking a transpose
matrix9.T

array([[1, 4, 7],
       [2, 5, 8],
       [3, 6, 9]])

**Function to find minimum and maximum values**

In [40]:
print(matrix9)

[[1 2 3]
 [4 5 6]
 [7 8 9]]


In [41]:
print('Minimum value: ',np.min(matrix9))

Minimum value:  1


In [42]:
print('Maximum value: ',np.max(matrix9))

Maximum value:  9


**Function to generate random samples**

**Using np.random.rand function**

* The np.random.rand returns a random NumPy array whose element(s) are drawn randomly from the uniform distribution over [0,1). (including 0 but excluding 1).
* **Syntax** - np.random.rand(d0,d1)
  * d0,d1 – It represents the dimension of the required array given as int, where d1 is optional.

In [43]:
# Generating random values in an array
rand_mat = np.random.rand(5)
print(rand_mat)

[0.71525375 0.3696061  0.78269507 0.69505709 0.62252667]


In [44]:
# * Generating random values in a matrix
rand_mat = np.random.rand(5,5) # uniform random variable
print(rand_mat)

[[0.03891921 0.20068123 0.50589807 0.72701016 0.01451509]
 [0.26424433 0.70745494 0.51530825 0.26880119 0.07600244]
 [0.34728755 0.57238067 0.08394086 0.1684161  0.68035085]
 [0.97202342 0.23111632 0.40034789 0.4358836  0.81454257]
 [0.94108531 0.0231066  0.71681185 0.40085156 0.34538775]]


**Using np.random.randn function**

* The np.random.randn returns a random numpy array whose sample(s) are drawn randomly from the standard normal distribution (Mean as 0 and standard deviation as 1)

* **Syntax** - np.random.randn(d0,d1)
  * d0,d1 – It represents the dimension of the output, where d1 is optional.

In [45]:
# Generating random values in an array
rand_mat2 = np.random.randn(5)
print(rand_mat2)

[-0.74009912  0.52601242  0.61839818 -0.80191091  0.31577375]


In [48]:
# Generating random values in a matrix
rand_mat2 = np.random.randn(500,500)
print(rand_mat2)

[[-1.58796562  0.6929772  -0.78829911 ...  1.6286079   0.40814678
   1.14829002]
 [ 0.59848342  1.18252916 -1.45071599 ...  0.31413948 -0.48831048
   0.8331378 ]
 [-0.84026182 -0.29805479  0.85274545 ...  0.53378338  0.56328404
  -0.67693667]
 ...
 [ 0.14687459  0.69237388 -0.29115173 ...  0.17127929 -0.76304416
   0.95164998]
 [-0.44149767  0.45217447  1.04923831 ...  0.14213757 -0.08584717
  -1.34524705]
 [ 0.45413081 -0.19727993 -2.49059504 ... -0.39353128 -0.01792763
   0.62568554]]


In [49]:
# Let's check the mean and standard deviation of rand_mat2
print('Mean:',np.mean(rand_mat2))
print('Standard Deviation:',np.std(rand_mat2))

Mean: -0.001085781378914981
Standard Deviation: 1.0007467176926834


*  We observe that the mean is very close to 0 and standard deviation is very close to 1.

**Using np.random.randint function**

* The np.random.randint returns a random numpy array whose element(s) are drawn randomly from low (inclusive) to the high (exclusive) range.

* **Syntax** - np.random.randint(low, high, size)

  * low – It represents the lowest inclusive bound of the distribution from where the sample can be drawn.
  * high – It represents the upper exclusive bound of the distribution from where the sample can be drawn.
  * size – It represents the shape of the output.

In [50]:
# Generating random values in an array
rand_mat3 = np.random.randint(1,5,10)
print(rand_mat3)

[2 2 3 4 4 3 2 1 2 4]


In [51]:
# Generating random values in a matrix
rand_mat3 = np.random.randint(1,10,[5,5])
print(rand_mat3)

[[8 9 6 3 5]
 [3 7 8 5 3]
 [7 4 2 8 8]
 [1 9 7 9 4]
 [8 9 7 4 7]]


### 2.3 Accessing the entries of a Numpy Array

In [52]:
# let's generate an array with 10 random values
rand_arr = np.random.randn(10)
print(rand_arr)

[ 0.0541112  -0.93100103 -1.25470322 -0.70664153 -0.16780983 -0.21389553
  0.56682038  0.20041157 -0.6655887   1.89207374]


* Accessing one element from an array

In [53]:
# accessing the 6 th entry of rand_arr
print(rand_arr[6])

0.5668203806936317


* Accessing multiple elements from an array

In [54]:
# we can access multiple entries at once using
print(rand_arr[4:9])

[-0.16780983 -0.21389553  0.56682038  0.20041157 -0.6655887 ]


In [55]:
# we can also access multiple non-consecutive entries using np.arange
print('Index of values to access: ',np.arange(3,10,3))
print(rand_arr[np.arange(3,10,3)])

Index of values to access:  [3 6 9]
[-0.70664153  0.56682038  1.89207374]


**Accessing arrays using logical operations**

In [56]:
print(rand_arr)

[ 0.0541112  -0.93100103 -1.25470322 -0.70664153 -0.16780983 -0.21389553
  0.56682038  0.20041157 -0.6655887   1.89207374]


In [57]:
rand_arr>0

array([ True, False, False, False, False, False,  True,  True, False,
        True])

In [58]:
# accessing all the values of rand_arr which are greater than 0
print('Values greater than 0: ',rand_arr[rand_arr>0])

# accessing all the values of rand_arr which are less than 0
print('Values less than 0: ',rand_arr[rand_arr<0])

Values greater than 0:  [0.0541112  0.56682038 0.20041157 1.89207374]
Values less than 0:  [-0.93100103 -1.25470322 -0.70664153 -0.16780983 -0.21389553 -0.6655887 ]


**Accessing the entries of a Matrix**

In [59]:
# let's generate an array with 10 random values
rand_mat = np.random.randn(5,5)
print(rand_mat)

[[ 0.7365679   1.20572771 -0.27417193  0.55128836  0.50145368]
 [ 1.44796851  0.48928197  1.10322374  0.7870271   0.6523692 ]
 [ 1.50274112  1.39908491 -0.92526491  1.654326   -0.04793924]
 [ 0.23829683  0.30965627  0.92115839 -2.55232906  1.39855443]
 [-0.33854558 -0.11921457  0.84890047  0.4577689   0.48801832]]


In [ ]:
# acessing the second row of the rand_mat
rand_mat[1]

array([ 0.68692112,  1.20042327, -0.55648032,  0.51675983, -1.13325252])

In [ ]:
# acessing third element of the second row
print(rand_mat[1][2])

#or

print(rand_mat[1,2])

-0.5564803167132045
-0.5564803167132045


In [ ]:
# accessing first two rows with second and third column
print(rand_mat[0:2,1:3])

[[-0.91860835 -0.91768744]
 [ 1.20042327 -0.55648032]]


**Accessing matrices using logical operations**

In [ ]:
print(rand_mat)

[[ 0.66008467 -0.91860835 -0.91768744 -0.56565779 -0.19013084]
 [ 0.68692112  1.20042327 -0.55648032  0.51675983 -1.13325252]
 [ 1.76883638 -0.85498119  1.89998922 -1.0658905   1.50531893]
 [ 0.77388464 -1.24206763  0.75631041  1.01940576  0.89747448]
 [ 0.37852758 -1.36846366 -0.91566544 -0.47943544 -0.20226054]]


In [ ]:
# accessing all the values of rand_mat which are greater than 0
print('Values greater than 0: \n ',rand_mat[rand_mat>0])

# accessing all the values of rand_mat which are less than 0
print('Values less than 0: \n',rand_mat[rand_mat<0])

Values greater than 0: 
  [0.66008467 0.68692112 1.20042327 0.51675983 1.76883638 1.89998922
 1.50531893 0.77388464 0.75631041 1.01940576 0.89747448 0.37852758]
Values less than 0: 
 [-0.91860835 -0.91768744 -0.56565779 -0.19013084 -0.55648032 -1.13325252
 -0.85498119 -1.0658905  -1.24206763 -1.36846366 -0.91566544 -0.47943544
 -0.20226054]


**Modifying the entries of an Array**

In [ ]:
print(rand_arr)

[-0.61610084  0.256255   -1.02474372  1.69663775  0.14830608  0.0587465
 -0.58451129  0.09827507  1.38926594  0.52114238]


In [ ]:
# let's change some values in an array!
# changing the values of index value 3 and index value 4 to 5
rand_arr[3:5] = 5
print(rand_arr)

[-0.61610084  0.256255   -1.02474372  5.          5.          0.0587465
 -0.58451129  0.09827507  1.38926594  0.52114238]


In [ ]:
# changing the values of index value 0 and index value 1 to 2 and 3 respectively
rand_arr[0:2] = [2,3]
print(rand_arr)

[ 2.          3.         -1.02474372  5.          5.          0.0587465
 -0.58451129  0.09827507  1.38926594  0.52114238]


In [ ]:
# modify entries using logical references
rand_arr[rand_arr>0] = 65
rand_arr

array([65.        , 65.        , -1.02474372, 65.        , 65.        ,
       65.        , -0.58451129, 65.        , 65.        , 65.        ])

**Modifying the entries of a Matrix**

In [60]:
print(rand_mat3)

[[8 9 6 3 5]
 [3 7 8 5 3]
 [7 4 2 8 8]
 [1 9 7 9 4]
 [8 9 7 4 7]]


In [61]:
# changing the values of the 4th and 5th element of the second and third rows of the matrix to 0
print('Matrix before modification: \n',rand_mat3)
rand_mat3[1:3,3:5] = 0
print('Matrix after modification: \n',rand_mat3)

Matrix before modification: 
 [[8 9 6 3 5]
 [3 7 8 5 3]
 [7 4 2 8 8]
 [1 9 7 9 4]
 [8 9 7 4 7]]
Matrix after modification: 
 [[8 9 6 3 5]
 [3 7 8 0 0]
 [7 4 2 0 0]
 [1 9 7 9 4]
 [8 9 7 4 7]]


In [62]:
# extracting the first 2 rows and first 3 columns from the matrix
sub_mat = rand_mat[0:2,0:3]
print(sub_mat)

[[ 0.7365679   1.20572771 -0.27417193]
 [ 1.44796851  0.48928197  1.10322374]]


In [63]:
# changing all the values of the extracted matrix to 3
sub_mat[:] = 3
print(sub_mat)

[[3. 3. 3.]
 [3. 3. 3.]]


In [ ]:
# what happened to rand_mat when we change sub_mat?
rand_mat

array([[ 3.        ,  3.        ,  3.        , -0.56565779, -0.19013084],
       [ 3.        ,  3.        ,  3.        ,  0.51675983, -1.13325252],
       [ 1.76883638, -0.85498119,  1.89998922, -1.0658905 ,  1.50531893],
       [ 0.77388464, -1.24206763,  0.75631041,  1.01940576,  0.89747448],
       [ 0.37852758, -1.36846366, -0.91566544, -0.47943544, -0.20226054]])

In [64]:
# to prevent this behavior we need to use the .copy() method when we assign sub_mat
# this behavior is the source of MANY errors for early python users!!!

rand_mat = np.random.randn(5,5)
print(rand_mat)
sub_mat = rand_mat[0:2,0:3].copy()
sub_mat[:] = 3
print(sub_mat)
print(rand_mat)

[[-0.45342124  0.44329407  0.61521361  0.29811461 -1.5060954 ]
 [ 0.46593571  0.43362975 -0.07393443 -0.25060543 -0.31965387]
 [ 0.25010803  0.30862983  0.95581027 -1.69665493 -0.41485968]
 [-1.51065554 -1.11450724 -2.128474   -0.33081207 -0.31801889]
 [ 0.7593488  -0.12237914  1.04246493 -0.39944956  1.09129567]]
[[3. 3. 3.]
 [3. 3. 3.]]
[[-0.45342124  0.44329407  0.61521361  0.29811461 -1.5060954 ]
 [ 0.46593571  0.43362975 -0.07393443 -0.25060543 -0.31965387]
 [ 0.25010803  0.30862983  0.95581027 -1.69665493 -0.41485968]
 [-1.51065554 -1.11450724 -2.128474   -0.33081207 -0.31801889]
 [ 0.7593488  -0.12237914  1.04246493 -0.39944956  1.09129567]]


### 2.4 Saving and Loading a NumPy array

**Let's save some NumPy objects on the disk for use later!**

In [65]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [66]:
# creating a random matrices
randint_matrix1 = np.random.randint(1,10,10).reshape(2,5)
print(randint_matrix1)
print('')
randint_matrix2 = np.random.randint(10,20,10).reshape(2,5)
print(randint_matrix2)

[[9 7 3 4 6]
 [8 4 3 3 4]]

[[16 19 15 19 19]
 [10 16 10 10 12]]


**Using np.save() function**

In [67]:
np.save('/content/drive/MyDrive/Python Course/saved_file_name',randint_matrix1)

**Using np.savez() function**

In [68]:
np.savez('/content/drive/MyDrive/Python Course/multiple_files',randint_matrix1=randint_matrix1,randint_matrix2=randint_matrix2)

* The files will be saved in the directory where the Jupyter Notebook is located.
* With np.save() function, we can save an array/matrix to a NumPy .npy format.
* np.savez() function has an advantage over np.save() function because with np.savez(), we can store several arrays/matrices into a single file in uncompressed .npz format.

In [69]:
# now let's load it
loaded_arr = np.load('/content/drive/MyDrive/Python Course/saved_file_name.npy')
loaded_multi = np.load('/content/drive/MyDrive/Python Course/multiple_files.npz')

print(loaded_arr)
print('')
print(loaded_multi)

[[9 7 3 4 6]
 [8 4 3 3 4]]

NpzFile '/content/drive/MyDrive/Python Course/multiple_files.npz' with keys: randint_matrix1, randint_matrix2


* We see that .npy file has been loaded but the .npz file is returning a memory location.
* Let's see how to load the values stored in .npz file.

In [70]:
print('1st Matrix: \n',loaded_multi['randint_matrix1'])
print('2nd Matrix: \n',loaded_multi['randint_matrix2'])

new_matrix  = loaded_multi['randint_matrix1']
print('New Matrix: \n',new_matrix)

1st Matrix: 
 [[9 7 3 4 6]
 [8 4 3 3 4]]
2nd Matrix: 
 [[16 19 15 19 19]
 [10 16 10 10 12]]
New Matrix: 
 [[9 7 3 4 6]
 [8 4 3 3 4]]


In [71]:
# we can also save/load text files...but only single variables
np.savetxt('/content/drive/MyDrive/Python Course/text_file_name.txt',randint_matrix1,delimiter=',')
rand_mat_txt = np.loadtxt('/content/drive/MyDrive/Python Course/text_file_name.txt',delimiter=',')
print(randint_matrix1)
print('')
print(rand_mat_txt)

[[9 7 3 4 6]
 [8 4 3 3 4]]

[[9. 7. 3. 4. 6.]
 [8. 4. 3. 3. 4.]]
